python-checks/src/main/java/org/sonar/python/checks/JwtVerificationCheck.java

```diff
@@ -38,39 +46,122 @@ public class JwtVerificationCheck extends PythonSubscriptionCheck {
+    "jose.jws.get_unverified_headers",
+    "jose.jwt.get_unverified_claims",
+    "jose.jws.get_unverified_claims"
+  );
+
+  private static final String VERIFY_SIGNATURE_KEYWORD = "verify_signature";
+
+  public static final String JOSE_JWT_DECODE_FQN = "jose.jwt.decode";
 
   @Override
   public void initialize(Context context) {
-    context.registerSyntaxNodeConsumer(Kind.CALL_EXPR, ctx -> {
-      CallExpression call = (CallExpression) ctx.syntaxNode();
-      Symbol calleeSymbol = call.calleeSymbol();
-      if (calleeSymbol == null) {
-        return;
+    context.registerSyntaxNodeConsumer(Kind.CALL_EXPR, JwtVerificationCheck::verifyCallExpression);
+  }
+
+  private static void verifyCallExpression(SubscriptionContext ctx) {
+    CallExpression call = (CallExpression) ctx.syntaxNode();
+    Symbol calleeSymbol = call.calleeSymbol();
+    if (calleeSymbol == null || calleeSymbol.fullyQualifiedName() == null) {
+      return;
+    }
+    if (WHERE_VERIFY_KWARG_SHOULD_BE_TRUE_FQNS.contains(calleeSymbol.fullyQualifiedName())) {
+      RegularArgument verifyArg = TreeUtils.argumentByKeyword("verify", call.arguments());
+      if (verifyArg != null && Expressions.isFalsy(verifyArg.expression())) {
+        ctx.addIssue(verifyArg, MESSAGE);
+      }
+    } else if (PROCESS_JWT_FQNS.contains(calleeSymbol.fullyQualifiedName())) {
+      Tree scriptOrFunction = TreeUtils.firstAncestorOfKind(call, Kind.FILE_INPUT, Kind.FUNCDEF);
+      if (!TreeUtils.hasDescendant(scriptOrFunction, JwtVerificationCheck::isCallToVerifyJwt)) {
+        ctx.addIssue(call, MESSAGE);
       }
-      if ("jwt.decode".equals(calleeSymbol.fullyQualifiedName())) {
-        RegularArgument verifyArg = TreeUtils.argumentByKeyword("verify", call.arguments());
-        if (verifyArg != null && Expressions.isFalsy(verifyArg.expression())) {
-          ctx.addIssue(verifyArg, MESSAGE);
-        }
-      } else if (PROCESS_JWT_NAMES.contains(calleeSymbol.fullyQualifiedName())) {
-        Tree scriptOrFunction = TreeUtils.firstAncestorOfKind(call, Kind.FILE_INPUT, Kind.FUNCDEF);
-        if (!TreeUtils.hasDescendant(scriptOrFunction, JwtVerificationCheck::isCallToVerifyJwt)) {
-          ctx.addIssue(call, MESSAGE);
-        }
+    } else if (UNVERIFIED_FQNS.contains(calleeSymbol.fullyQualifiedName())) {
+      ctx.addIssue(call, MESSAGE);
+    } else if (JOSE_JWT_DECODE_FQN.equals(calleeSymbol.fullyQualifiedName())) {
+      RegularArgument optionsArg = TreeUtils.argumentByKeyword("options", call.arguments());
+      if (optionsArg != null && isListOrDictWithSensitiveEntry(optionsArg.expression())) {
+        ctx.addIssue(optionsArg, MESSAGE);
       }
-    });
+    }
+  }
+
+  private static boolean isListOrDictWithSensitiveEntry(@Nullable Expression expression) {
+    if (expression == null) {
+      return false;
+    } else if (expression.is(Tree.Kind.NAME)) {
+      return isListOrDictWithSensitiveEntry(Expressions.singleAssignedValue((Name) expression));
```

Comment:
Remembering the StackOverflowException we investigated recently, could this trigger one as well in presence of circular assignments? As far as I remember, we didn't fix the underlying cause in `Expression#singleAssignedValue`?

Reply:
Yes, indeed I am trying to resolve variables in the same way. I have added the test case that causes the infinite recursion in 46c635a1b0b8eefaef0a39bb623382e0ae0f577e. However, in this case, it doesn't seem to break, and I am not entirely sure why. We can discuss this maybe.

Reply:
In your test, the recursion is not triggered because it's only done in the `jose.jwt.decode` case. The example uses `jose.jws.verify` for which a simple falsy check on the `verify` argument value is done.

Reply:
This case will trigger an infinite recursion:
```
x = []
class TestInfiniteRecursion():
    x = x
    payload = jwt.decode(token, None, options=x)
```